nightfall-optimist/src/services/block-assembler.mjs

```diff
@@ -62,6 +63,12 @@ async function makeBlock(proposer, number = TRANSACTIONS_PER_BLOCK) {
   // we retrieve un-processed transactions from our local database, relying on
   // the transaction service to keep the database current
   const transactions = await getMostProfitableTransactions(number);
+
+  // check transction before building block
+  transactions.forEach(async transaction => {
+    await checkTransaction(transaction, true);
```

Comment:
Why are you setting checkTransaction parameter inL2AndNotInL2 to true?

Reply:
In fact, couldn't we remove this parameter? Since duplicate commitments nullifiers is already checked in the circuit, this is no longer needed. If a transaction have a duplicate nullifier / commitment in the same tx the proof verification will fail

Reply:
> Why are you setting checkTransaction parameter inL2AndNotInL2 to true?
It needs to be `false`.

`In fact, couldn't we remove this parameter? Since duplicate commitments nullifiers is already checked in the circuit, this is no longer needed. If a transaction have a duplicate nullifier / commitment in the same tx the proof verification will fail`
The logic with duplicate commitment or nullifier will get more complicated than it is today. 
For example, during transactionSubmittedEventHandler, if there is a duplicate commitment from a new transaction, it can still be added into the mempool as long as the fee paid for this transaction is higher and the commitment already in DB is not part of L2 block yet. Also, if the commitment already in DB is part of an L2 block but it is part of an invalid transaction, then it this new transaction if it is valid, should be added to DB because this transaction will be proposed after rollback removes the commitment in DB. This logic will be handled in a different issue but since `checkTransaction` will be used at transactionSubmittedEventHandler, blockAssembler, rollbackEventHandler, blockProposedEventHandler, it needs to handle error handling correctly based on flags such as - `isAlreadyInL2` and `isAlreadyInMempool`

Similar complex handling will exist at blockProposed handler stage too. If the commitment from a transaction from block proposed stage is a duplicate of another commitment already in L2 block, then raise a challenge. Instead if it is a duplicate of another commitment already in mempool and not in L2 block, then mark this mempool commitment which is part of a different transaction to `false` too.

I will create a separate issue for this.

Reply:
Meanwhile @LijuJoseJJ please make the following changes.

Replacing `inL2AndNotInL2` with more apt and clear flags of  `isAlreadyInL2` and `isAlreadyInMempool`. For the moment, if a new commitment or nullifier received already exists in L2  or mempool during `TransactionSubmittedEventHandler`, we want to detect and not save this new transaction received. For BlockProposedEventHandler, RollbackEventHandler, BlockAssembly, we only care if the new commitment or nullifier received is in L2 and handle according to the context. The following changes should help

- Replace, `inL2AndNotInL2` (terrible and confusing name) with `isAlreadyInL2` and `isAlreadyInMempool` in `checkTranasction` 
- Flag values
  - BlockProposedEventHandler
    - isAlreadyInL2: true
    - isAlreadyInMempool: false
  - RollbackEventHandler
    - isAlreadyInL2: true
    - isAlreadyInMempool: false
  - TransactionSubmittedEventHandler
    - isAlreadyInL2: true
    - isAlreadyInMempool: true
  - BlockAssembly
    - isAlreadyInL2: true
    - isAlreadyInMempool: false
 - Replace  ` const query = inL2AndNotInL2
    ? { nullifiers: nullifierHash }
    : {
        nullifiers: nullifierHash,
        blockNumberL2: { $gte: -1, $ne: blockNumberL2OfTx },
      };` in `getL2TransactionByNullifier` with  `const query = (isAlreadyInL2 && isAlreadyInMempool)
    ? { nullifiers: nullifierHash }
    : {
        nullifiers: nullifierHash,
        blockNumberL2: { $gte: -1, $ne: blockNumberL2OfTx },
      };`
 - Replace   `const query = inL2AndNotInL2
    ? { commitments: commitmentHash }
    : {
        commitments: commitmentHash,
        blockNumberL2: { $gte: -1, $ne: blockNumberL2OfTx },
      };` in `getL2TransactionByCommitment` with `const query = (isAlreadyInL2 && isAlreadyInMempool)
    ? { commitments: commitmentHash }
    : {
        commitments: commitmentHash,
        blockNumberL2: { $gte: -1, $ne: blockNumberL2OfTx },
      };`
 - replace `inL2AndNotInL2` in `TransactionError` in `checkDuplicateCommitment` and `checkDuplicateNullifier` with `isAlreadyInL2`

Reply:
done